File: clients/python/src/munarium_client/_retry.py
# SPDX-License-Identifier: Apache-2.0
"""Retry pacing: jittered exponential backoff (clock-nanos jitter — no
security requirement, just pacing)."""

from __future__ import annotations

import asyncio
import time


def backoff_seconds(attempt: int) -> float:
    """Jittered backoff for attempt n (1-based): base 2^(n-1) * 100ms, ±50%,
    clamped to [0.05s, 2s]."""
    base_ms = 100 * (1 << min(attempt - 1, 6))
    jitter = time.time_ns() % max(base_ms, 1)
    return min(max((base_ms / 2 + jitter) / 1000.0, 0.05), 2.0)


def sleep_sync(attempt: int) -> None:
    time.sleep(backoff_seconds(attempt))


async def sleep_async(attempt: int) -> None:
    await asyncio.sleep(backoff_seconds(attempt))

File: clients/rust/munarium-client/src/retry.rs
// SPDX-License-Identifier: Apache-2.0
//! Retry pacing: jittered exponential backoff without a rand dependency
//! (subsecond clock nanos are jitter enough for pacing).

use std::time::Duration;

/// Jittered backoff for attempt n (1-based): base 2^(n-1) * 100ms, ±50%,
/// clamped to [50ms, 2s].
pub(crate) async fn jitter_sleep(attempt: u32) {
    let base_ms = 100u64.saturating_mul(1 << attempt.saturating_sub(1).min(6));
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.subsec_nanos() as u64)
        .unwrap_or(0);
    let jitter = nanos % base_ms.max(1);
    let ms = (base_ms / 2 + jitter).clamp(50, 2_000);
    tokio::time::sleep(Duration::from_millis(ms)).await;
}
